### DISCUSSION TITLE
Find And Replace in yml files

### DISCUSSION THREAD

**Discussion by dayagz** [openrewrite/rewrite#4550] [Ideas]
OpenRewrite YAML recipes currently do not support find and replace operations. The text-based find and replace modifies the structure, which prevents subsequent recipes from being applied. Could we consider adding a find and replace recipe?

**Comment by shanman190** [comment]
Find and replace is conceptually an activity applied to textual source. When using the `org.openrewrite.text.FindAndReplace`, it converts the current source file to its textual form and then applies the recipe over it. As a result, the original LST model is lost and instead a `PlainText` LST element is returned in its place.

Using find and replace in a declarative YAML recipe will work, but it's a very common mistake that users run into when they then try to then compose language aware recipes _after_ the find and replace operation occurred. So often a mistake that it's documented explicitly that language specific operations should be applied first and then find and replace applied at the very end. You can see that documented here: https://docs.openrewrite.org/recipes/text/findandreplace

**Reply by knutwannheden** [reply]
It appears to me that if we were to offer a parsing service (e.g. by pulling up the JavaSourceFile#service() to SourceFile), then a language that supports "reparse" could allow recipes like FindAndReplace to "rehydrate" an LST from text. Parsers like Java won't be able to support this for the foreseeable future, due to the type attribution, but for YAML this should not be an issue.

Maybe @timtebeek has seen similar requests or applications.

**Reply by shanman190** [reply]
@knutwannheden, that's fair. Assuming "reparsing" is what's being asked for that would probably be limited to just hcl, json, properties, and yaml. As all of the others have type attribution or other complex scenarios.

**Reply by knutwannheden** [reply]
Correct. And XML (except for Maven pom.xml files of course, unless losing the `MavenResolutionResult` marker is considered acceptable).

Possibly I am just looking for a good excuse to pull up that `service()` method 😅

**Reply by dayagz** [reply]
That would be a nice feature to add, as these config files often require a value to be replaced first and then continue with language-specific operations, such as removing an empty key that results from the replace operation. The ability to rehydrate the LST after a find and replace would solve this issue and allow for a seamless workflow where modifications to the text are followed by structural changes.

**Comment by knutwannheden** [comment]
@dayagz Do you not know what mapping needs replacing in your YAML? If you can't use ChangePropertyValue you could also implement a custom recipe that replaces any scalar value matching a given substring or pattern.

**Reply by dayagz** [reply]
@knutwannheden i had to change a value that matches a pattern, and created a custom recipe since i was using text based find and replace altered the structure i couldn't run any subsequent recipes. I believe had a native find and replace capability in in the yml will be helpful as lot of application configurations are defined in yamls

**Reply by timtebeek** [reply]
There's various Yaml specific recipes that allow you to change keys, properties & values: https://docs.openrewrite.org/recipes/yaml#recipes
What would you need that goes beyond what's available already?

**Comment by chris-nielsen-wgu** [comment]
Our organization would also love to see this feature: Presently, we've started creating a recipe to migrate from Micronaut 4 to Spring Boot 3.x. (Ideally we would use an existing one, but plenty of Googling suggests this doesn't exist at the moment...?)

And one of the bigger pain points, is migrating Yaml-based config.

Ideally we would like to be able to use wildcard-based replacements, when properties map 1:1 between the two. DB connection info is one place where wildcards might work.

To be fair, we likely _will_ need quite a few individual replacements, given the differences between the two frameworks. But for the specific cases where we _can_ use a wildcard [for key name replacement], that would be tremendously helpful!

Questions, thoughts, suggestions...?
